Sources/NIO/LinuxUring.swift

```diff
@@ -0,0 +1,509 @@
+    private let cqeMaxCount : UInt32 = 8192 // this is the max chunk of CQE we take.
+
+    var cqes : UnsafeMutablePointer<UnsafeMutablePointer<io_uring_cqe>?>
+    var fdEvents = [fdEventKey : UInt32]() // fd, sequence_identifier : merged event_poll_return
+    var emptyCqe = io_uring_cqe()
+
+    internal static let initializedUring: Bool = {
+        CNIOLinux_check_compatibility() == 0
+    }()
+
+    internal func fd() -> Int32 {
+       return ring.ring_fd
+    }
+
+    func _dumpCqes(_ header:String, count: Int = 1)
+    {
+        #if SWIFTNIO_LIBURING_DEBUG_DUMP_CQE
+        func _debugPrintCQE(_ s : String) {
+            print("Q [\(NIOThread.current)] " + s)
+        }
+
+        if count < 0 {
+            return
+        }
+
+        _debugPrintCQE(header + " CQE:s [\(cqes)] - ring flags are [\(ring.flags)]")
+        for i in 0..<count {
+            let c = cqes[i]!.pointee
+
+            let dp = io_uring_cqe_get_data(cqes[i])
+            let bitPattern : UInt = UInt(bitPattern:dp)
+            let fd = Int32(bitPattern & 0x00000000FFFFFFFF)
+            let sequenceNumber : UInt32 = UInt32((Int(bitPattern) >> 32) & 0x00FFFFFF)
+            let eventType = CqeEventType(rawValue:((Int(bitPattern) >> 32) & 0xFF000000) >> 24) // shift out the fd
+
+            let bitpatternAsPointer = UnsafeMutableRawPointer.init(bitPattern: bitPattern)
+
+            _debugPrintCQE("\(i) = fd[\(fd)] eventType[\(String(describing:eventType))] sequenceNumber[\(sequenceNumber)] res [\(c.res)] flags [\(c.flags)]  bitpattern[\(String(describing:bitpatternAsPointer))]")
+        }
+        #endif
+    }
+
+    init() {
+        cqes = UnsafeMutablePointer<UnsafeMutablePointer<io_uring_cqe>?>.allocate(capacity: Int(cqeMaxCount))
+        cqes.initialize(repeating:&emptyCqe, count:Int(cqeMaxCount))
+    }
+
+    deinit {
+        cqes.deallocate()
+    }
+
+    internal func io_uring_queue_init() throws -> () {
+        // IORING_SETUP_SQPOLL will be fundamentally useful in 5.13, as we no longer need elevated privileges as previously
+        // and CNIOLinux.io_uring_queue_init will setup a shared uring instance which allows us to use a single
+        // kernel sqpoll thread that is shared amongst all loops. NB, each process will have it's own polling thread
+        // if SQPOLL is enabled, so for hosts running multiple processes using it, some care should be taken to not overload.
+        #if SWIFTNIO_LIBURING_SQPOLL_ENABLED
+        let _sqpollEnabled : UInt32 = 1
+        #else
+        let _sqpollEnabled : UInt32 = 0
+        #endif
+        if (CNIOLinux_io_uring_queue_init(ringEntries, &ring, _sqpollEnabled ) != 0)
+         {
+             throw UringError.uringSetupFailure
+         }
+
+        _debugPrint("io_uring_queue_init \(self.ring.ring_fd)")
+     }
+
+    internal func io_uring_queue_exit() {
+        _debugPrint("io_uring_queue_exit \(self.ring.ring_fd)")
+        CNIOLinux.io_uring_queue_exit(&ring)
```

Comment:
can all these `io_uring_*` calls not return `EINTR`?

Reply:
There are fundamentally three io_uring system calls that can result from using the liburing functions - io_uring_setup, io_uring_register and io_uring_enter - only the last can return EINTR. That only occur according to documentation in “Can happen while waiting for events with IORING_ENTER_GETEVENTS” which is basically fine (that is when we are blocked waiting for events, so we would just return no events and run the next event loop tick).

Reply:
(most of the io_uring helper functions just set up fields and prepare for calling one of those syscalls in a nicer way, except for queue setup/exit that also `mmap()`, but that can't EINTR AFAIU and would be a bug in the liburing library if it did.)